src/components/application_manager/src/state_controller_impl.cc

```diff
@@ -245,129 +245,185 @@ void StateControllerImpl::SetRegularState(ApplicationSharedPtr app,
-      state_->hmi_level() == mobile_apis::HMILevel::HMI_FULL;
-
-  const bool applied_grabs_audio =
-      HasStreamingPermissions(state_->hmi_level()) &&
-      applied_->IsAudioApplication();
+  // {
+  //   const std::string nl_indent = "\n ---> ";
+  //   std::stringstream ss;
+  //   ss << '\n';
+  //   ss <<
+  //   "-------------------HMI_CONFLICT_RESOLVER-------------------";
+  //   ss << nl_indent << "Applied app id    - " << applied_->app_id();
+  //   ss << nl_indent << "Applied reg state - " << *state_;
+  //   ss << nl_indent << "Resolve app id    - " << app_to_resolve->app_id();
+  //   ss << nl_indent << "Resolve reg state - " << *state_to_resolve;
+  //   LOG4CXX_DEBUG(logger_, ss.str());
+  // }
+
+  // If applied HMI state is FULL:
+  // - all NOT audio/video applications becomes BACKGROUND
+  // - all audio/video applications with other app type
+  //  (navi, vc, media, projection) in FULL becomes LIMMITED
+  // - all applications with same app type becomes BACKGROUND
+  //
+  // If applied HMI state is LIMITED:
+  // - all NOT audio/video applications saves their's HMI states
+  // - all applications with other app types saves their's HMI states
+  // - all applications with same app type becomes BACKGROUND
+  //
+  // If applied HMI state is BACKGROUND:
+  // - all applications will save their's HMI states
+  //
+  // Resolve tables to figure out audio/video streaming state
+  // for FULL/LIMITED HMI level.
+  // For example if applied is 'NAVIGATION' and resolve app is
+  // 'PROJECTION' then to resolve becomes 'NOT_STREAMABLE'.
+  // According to video resolve table the only one app may
+  // be 'STREAMABLE' in the same time.
+  //
+  // Video app resolve table
+  // * S --> STREAMABLE, NS --> NOT_STREAMABLE
+  // | resolve / applied -->
+  // V
+  // _______navi___proj__other__
+  // navi  |__NS__|_NS__|__S____|
+  // proj  |__NS__|_NS__|__S ___|
+  // other |__NS__|_NS__|__NS___|
+  //
+  // Audio app resolve table
+  // * A --> AUDIBLE, NS --> NOT_AUDIBLE
+  // | resolve / applied -->
+  // V
+  // _______navi___vc___media__other__
+  // navi  |_NA__|_A___|__A__|____A___|
+  // vc    |__A__|_NA__|__A__|____A___|
+  // media |__A__|_A___|__NA_|____A___|
+  // other |_NA__|_NA__|__NA_|____NA__|
+  //
+  // NOTE: for instance if applied app is 'PROJECTION' & isMedia=true
+  // with applied HMI level 'FULL' (STREAMABLE & AUDIBLE)
+  // and app to resolve is 'NAVIGATION' with HMI level FULL or LIMITED
+  // (STREAMABLE & AUDIBLE) then to resolve becomes LIMITED
+  // (NOT_STREAMABLE & AUDIBLE). So, the only one app may stream video
+  // in the same time. But audio app such types as
+  // NAVIGATION, COMMUNICATION may stream audio/video in LIMITED level
+  // with app of other types.
+  // NOTE: if applied app is 'NAVIGATION'
+  // (STREAMABLE & AUDIBLE) and app to resolve is same type
+  // 'NAVIGATION' (STREAMABLE & AUDIBLE) then to resolve
+  // becomes BACKGROUND (NOT_STREAMABLE & NOT_AUDIBLE).
+
```

Comment:
@EKuliiev Please don't put commented out blocks of code in repository.
Debug code separate out into a separate function.
For description make topic in confluence and just put the link.